accepted/2021/json-polymorphism.md

```diff
@@ -0,0 +1,275 @@
+In the above example `Bar` inherits from `Foo` yet is not polymorphic in its own right:
+```csharp
+Bar bar = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Bar>(bar); // { "A" : 1, "B" : 2 }
+```
+If annotating the base class with an attribute is not possible,
+polymorphism can alternatively be opted in for a type using the
+new `JsonSerializerOptions.SupportedPolymorphicTypes` predicate:
+```csharp
+public class JsonSerializerOptions
+{
+  public Func<Type, bool> SupportedPolymorphicTypes { get; set; }
+}
+```
+Applied to the example above:
+```csharp
+var options = new JsonSerializerOptions { SupportedPolymorphicTypes = type => type == typeof(Foo) };
+JsonSerializer.Serialize<Foo>(foo1, options); // { "A" : 1, "B" : 2 }
+JsonSerializer.Serialize<Foo>(foo2, options); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+It is always possible to use this setting to enable polymorphism _for every_ serialized type:
+```csharp
+var options = new JsonSerializerOptions { SupportedPolymorphicTypes = _ => true };
+
+// `options` treats both `Foo` and `Bar` members as polymorphic
+Baz baz = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Foo>(baz, options); // { "A" : 1, "B" : 2, "C" : 3 }
+JsonSerializer.Serialize<Bar>(baz, options); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+As mentioned previously, this feature provides no provision for deserialization.
+If deserialization is a requirement, users would need to opt for the
+polymorphic serialization with type discriminators feature.
+
+## Polymorphism with type discriminators
+
+This feature allows users to opt in to polymorphic serialization for a given type
+by associating string identifiers with particular subtypes in the hierarchy.
+These identifiers are written to the wire so this brand of polymorphism is roundtrippable.
+
+At the core of the design is the introduction of `JsonKnownType` attribute that can
+be applied to type hierarchies like so:
+```csharp
+[JsonKnownType(typeof(Derived1), "derived1")]
+[JsonKnownType(typeof(Derived2), "derived2")]
+public class Base
+{
+    public int X { get; set; }
+}
+
+public class Derived1 : Base
+{
+    public int Y { get; set; }
+}
+
+public class Derived2 : Base
+{
+    public int Z { get; set; }
+}
+```
+This allows roundtrippable polymorphic serialization using the following schema:
+```csharp
+var json1 = JsonSerializer.Serialize<Base>(new Derived1()); // { "$type" : "derived1", "X" : 0, "Y" : 0 }
+var json2 = JsonSerializer.Serialize<Base>(new Derived2()); // { "$type" : "derived2", "X" : 0, "Z" : 0 }
+
+JsonSerializer.Deserialize<Base>(json1); // uses Derived1 as runtime type
+JsonSerializer.Deserialize<Base>(json2); // uses Derived2 as runtime type
+```
+Alternatively, users can specify known type configuration using the
+`JsonSerializerOptions.TypeDiscriminatorConfigurations` property:
+```csharp
+public class JsonSerializerOptions
+{
+    public IList<TypeDiscriminatorConfiguration> TypeDiscriminatorConfigurations { get; }
```

Comment:
I was thinking this should be a dictionary, but then I thought if we expose a way to get `JsonTypeInfo`, someone could set the list directly for the type. e.g.:
```cs
JsonTypeInfo typeinfo = options.GetTypeInfo<Base>();
IList<KnownType> knownTypes = typeinfo.KnownTypes;
knownTypes.Add<Derived1>("derived1");
knownTypes.Add<Derived2>("derived2");
```

Reply:
FWIW the configuration is used to construct [a class that is backed by a bidirectional map](https://github.com/dotnet/runtime/pull/53882/files#diff-ea8cdec34c5ef04cf0ce79e23e722bccb113c41251c53d36e4d7697dc39abf36).